mm2src/coins/hd_wallet_storage/sqlite_storage.rs

```diff
@@ -0,0 +1,300 @@
+    ) -> HDWalletStorageResult<()> {
+        self.update_addresses_number(
+            "external_addresses_number",
+            wallet_id,
+            account_id,
+            new_external_addresses_number,
+        )
+        .await
+    }
+
+    async fn update_internal_addresses_number(
+        &self,
+        wallet_id: HDWalletId,
+        account_id: u32,
+        new_internal_addresses_number: u32,
+    ) -> HDWalletStorageResult<()> {
+        self.update_addresses_number(
+            "internal_addresses_number",
+            wallet_id,
+            account_id,
+            new_internal_addresses_number,
+        )
+        .await
+    }
+
+    async fn upload_new_account(
+        &self,
+        wallet_id: HDWalletId,
+        account: HDAccountStorageItem,
+    ) -> HDWalletStorageResult<()> {
+        let selfi = self.clone();
+        async_blocking(move || {
+            let conn_shared = selfi.get_shared_conn()?;
+            let conn = Self::lock_conn(&conn_shared)?;
+
+            let params = account.to_sql_params_with_wallet_id(wallet_id);
+            conn.execute(INSERT_ACCOUNT, params)
+                .map(|_| ())
+                .map_to_mm(HDWalletStorageError::from)
+        })
+        .await
+    }
+
+    async fn clear_accounts(&self, wallet_id: HDWalletId) -> HDWalletStorageResult<()> {
+        let selfi = self.clone();
+        async_blocking(move || {
+            let conn_shared = selfi.get_shared_conn()?;
+            let conn = Self::lock_conn(&conn_shared)?;
+
+            let params = wallet_id.to_sql_params();
+            conn.execute(DELETE_ACCOUNTS_BY_WALLET_ID, params)
+                .map(|_| ())
+                .map_to_mm(HDWalletStorageError::from)
+        })
+        .await
+    }
+}
+
+impl HDWalletSqliteStorage {
+    fn get_shared_conn(&self) -> HDWalletStorageResult<SqliteConnShared> {
+        self.conn
+            .upgrade()
+            .or_mm_err(|| HDWalletStorageError::Internal("'HDWalletSqliteStorage::conn' doesn't exist".to_owned()))
+    }
+
+    fn lock_conn(conn: &SqliteConnShared) -> HDWalletStorageResult<MutexGuard<Connection>> {
+        conn.lock()
+            .map_to_mm(|e| HDWalletStorageError::Internal(format!("Error locking sqlite connection: {}", e)))
+    }
+
+    async fn init_tables(&self) -> HDWalletStorageResult<()> {
+        let conn_shared = self.get_shared_conn()?;
+        let conn = Self::lock_conn(&conn_shared)?;
+        conn.execute(CREATE_HD_ACCOUNT_TABLE, NO_PARAMS)
+            .map(|_| ())
+            .map_to_mm(HDWalletStorageError::from)
+    }
+
+    async fn update_addresses_number(
+        &self,
+        updating_field: &str,
+        wallet_id: HDWalletId,
+        account_id: u32,
+        new_addresses_number: u32,
+    ) -> HDWalletStorageResult<()> {
+        let sql = format!(
+            "UPDATE hd_account SET {}=?1 WHERE coin=?2 AND mm2_rmd160=?3 AND hd_wallet_rmd160=?4 AND account_id=?5;",
+            updating_field
```

Comment:
This opens a possibility for SQL injection. Please validate or sanitize this value before using in the query.

Reply:
Done